channel.go

```diff
@@ -0,0 +1,258 @@
+package stream_chat
+
+import (
+	"fmt"
+	"net/http"
+)
+
+const (
+	channelPathFmt = "channels/%s/%s"
+	messagePathFmt = "channels/%s/%s/message"
+	eventPathFmt   = "channels/%s/%s/event"
+)
+
+type Channel struct {
+	id         string
+	_type      string
+	client     *client
+	customData map[string]interface{}
+}
+
+func (ch Channel) formatPath(path string, params ...interface{}) string {
+	if ch.id == "" {
+		return ""
+	}
+
+	params = append([]interface{}{ch._type, ch.id}, params...)
+
+	return fmt.Sprintf(path, params)
+}
+
+func addUserID(hash map[string]interface{}, userID string) map[string]interface{} {
+	hash["user"] = userID
+	return hash
+}
+
+// SendMessage sends a message to this channel
```

Comment:
Message is not a bit struct. Basically there are no big structs in go (10 int64 fields == 80bytes), so putting on stack is faster and such kind of tuning as overusing pointers makes more damage than helps

Reply:
A heap copy of the 30 byte Message is the opposite of fast.

You are right that Go uses the stack a bit more than other languages when immutability is key and the struct is small. Examples in the core library are the 3 byte time.Time and reflect.Value.

Another good use case is heap allocation for byte arrays such as the return of sha1.Sum.

Reply:
```
pkg: github.com/GetStream/stream-chat-go
BenchmarkHeap-12       	222333336	         5.39 ns/op
BenchmarkPointer-12    	639435444	         1.88 ns/op
```

```go
package stream_chat

import "testing"

var sum int // prevents compiler optimisation

func Heap(m Message) {
        sum += m.ReplyCount
}

func Pointer(m *Message) {
        sum += m.ReplyCount
}

func BenchmarkHeap(b *testing.B) {
        var m Message
        for i := 0; i < b.N; i++ {
                Heap(m)
        }
}

func BenchmarkPointer(b *testing.B) {
        var m Message
        for i := 0; i < b.N; i++ {
                Pointer(&m)
        }
}
```

Reply:
Your test is correct if we are sending same message billion times.
But does we do that? Usually it's used in different way:

```
package stream_chat

import "testing"

var sum int // prevents compiler optimisation

func SendMessageHeap(m Message) {
	sum += m.ReplyCount
}

func SendMessagePointer(m *Message) {
	sum += m.ReplyCount
}

func BenchmarkHeap(b *testing.B) {
	b.ReportAllocs()

	for i := 0; i < b.N; i++ {
		var m = Message{ID: "test", ReplyCount: i}
		SendMessageHeap(m)
	}
}

func BenchmarkPointer(b *testing.B) {
	b.ReportAllocs()

	for i := 0; i < b.N; i++ {
		var m = &Message{ID: "test", ReplyCount: i}
		SendMessagePointer(m)
	}
}
```

```
pkg: github.com/GetStream/stream-chat-go
BenchmarkHeap-4         2000000000               1.55 ns/op            0 B/op          0 allocs/op
BenchmarkPointer-4      2000000000               1.68 ns/op            0 B/op          0 allocs/op
```

Reply:
You are right in that the overhead doesn't really matter that much, since the overhead is low. What I'm saying is that you shouldn't use struct copies as a default. There's specific cases where they add value, or even increase performance. Using struct copies can also introduce weird behaviour, which is hard to understand for people with less knowledge about the internal workings. So just use pointers, unless there's a reason not to. Not the other way around.

Reply:
gotcha. in any case I've found that message will be updated from response body, so we need pointer anyway